gems/sorbet-runtime/lib/types/props/private/setter_factory.rb

```diff
@@ -26,14 +26,20 @@ def self.build_setter_proc(klass, prop, rules)
         non_nil_type = T::Utils::Nilable.get_underlying_type_object(rules.fetch(:type_object))
         accessor_key = rules.fetch(:accessor_key)
         validate = rules[:setter_validate]
+        check_level = rules.fetch(:checked)
 
         # It seems like a bug that this affects the behavior of setters, but
         # some existing code relies on this behavior
         has_explicit_nil_default = rules.key?(:default) && rules.fetch(:default).nil?
+        requires_validation = check_level == :always || (check_level == :tests && T::Private::RuntimeLevels.check_tests?)
 
         # Use separate methods in order to ensure that we only close over necessary
         # variables
-        if !T::Props::Utils.need_nil_write_check?(rules) || has_explicit_nil_default
+        if !requires_validation
+          proc do |val|
+            instance_variable_set(accessor_key, val)
+          end
+        elsif !T::Props::Utils.need_nil_write_check?(rules) || has_explicit_nil_default
```

Comment:
I have a feeling that what we are going to want is that if this codepath would execute, we want to actually define the setter method at runtime with `attr_writer`. The Ruby VM special cases methods declared with `attr_writer` to run substantially faster than a method that calls `instance_variable_set`.

Reply:
We do something similar with `attr_reader` for a prop's getter method here:

https://github.com/sorbet/sorbet/blob/213e156fe200855e30781c1bab1e7fab5db88a72/gems/sorbet-runtime/lib/types/props/decorator.rb#L413-L414

Reply:
That wouldn’t work for the immutable case, right? We could potentially `undef` the writers at the end of the initialize method, if it’s a similar speed up maybe it would still be faster. I can try it out and report back.

I think @djudd’s idea from slack would probably be the ideal in terms of perf:
> Speaking of which though, a more ambitious perf thing would be to start generating constructors as Ruby code and eval-ing them like we do with serde. There's some meaningful overhead associated with the current abstractions for small/hot objects. Depends how comfortable we are with the eval stuff though

Reply:
Also thinking more about it now, if you use `attr_writer` you'll need to `@instance.send(:propname)` for each one in the initializer which is slower than `instance_variable_set(:@propname)`.

Reply:
Why does the constructor depend on the setter methods? They should be separate code paths.

Reply:
The `setter_proc` is currently deeply intertwined with constructor of T::Struct, here are some examples:
- `T::Props::WeakConstructor` and `T::Props::Constructor` work by iterating over the props and `instance_exec` the setter proc, for example: https://github.com/sorbet/sorbet/blob/f8835e7eb5067c5e3f0da8db5dcb0e7867c0bacd/gems/sorbet-runtime/lib/types/props/weak_constructor.rb#L36-L41
- `T::Props::ApplyDefault.for` (which is used for props with defaults) also uses `instance_exec` to validate the defaults, for example: https://github.com/sorbet/sorbet/blob/f8835e7eb5067c5e3f0da8db5dcb0e7867c0bacd/gems/sorbet-runtime/lib/types/props/private/apply_default.rb#L160

Reply:
@jez - I updated `T::Props::Decorator#define_getter_and_setter` to check if we need to typecheck the prop and use `attr_writer` if not. See the last commit and the updated PR description with the benchmarks. It basically gets us back to native PORO speed 🚀 

Let me know if you have thoughts on how to unwind the setter_proc from the constructors. If that winds up being a larger refactor I also wonder if we could consider merging this PR first and following up with that. Happy to help keep driving performance here since we rely heavily on T::Struct.